### PR TITLE
update identity in vmss only once

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

Update identity in vmss only once as identity is across all nodes in vmss and not just single node.

**Issue Fixed**:

Fixes https://github.com/Azure/aad-pod-identity/issues/361

**Notes for Reviewers**:

### CODE REVIEW

File: pkg/mic/vmss.go [hunk 1]
```diff
@@ -96,6 +96,11 @@ func makeVMSSID(r azure.Resource) string {
 	return path.Join(r.SubscriptionID, r.ResourceGroup, r.ResourceName)
 }
 
+func getVMSSName(vmssID string) string {
+	_, resourceName := path.Split(vmssID)
```
Comment (kkmsft) [inline pkg/mic/vmss.go]: Is it possible to have an erroneous vmssID, if so how do we handle this ?

File: pkg/mic/mic.go [hunk 2]
```diff
@@ -874,31 +865,72 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
...
 
-	for _, deleteID := range nodeTrackList.assignedIDsToDelete {
-		binding := deleteID.Spec.AzureBindingRef
+func (c *Client) consolidateVMSSNodes(nodeMap map[string]trackUserAssignedMSIIds) {
+	vmssMap := make(map[string][]string)
 
-		err := c.removeAssignedIdentity(&deleteID)
+	for nodeName, nodeTrackList := range nodeMap {
+		node, err := c.NodeClient.Get(nodeName)
+		if err != nil && !strings.Contains(err.Error(), "not found") {
```
Comment (kkmsft) [inline pkg/mic/mic.go]: const ?

File: pkg/mic/mic.go [hunk 3]
```diff
@@ -874,31 +865,72 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
...
+			glog.Errorf("Unable to get node %s. Error %v", nodeName, err)
+			continue
+		}
+		if err != nil && strings.Contains(err.Error(), "not found") {
+			glog.Warningf("Unable to get node %s while updating user msis. Error %v", nodeName, err)
```
Comment (kkmsft) [inline pkg/mic/mic.go]: spurious line ?

File: pkg/mic/mic.go [hunk 4]
```diff
@@ -874,31 +865,72 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
...
+
+			// node is no longer found in the cluster, all the assigned identities that were created in this sync loop
+			// and those that already exist for this node need to be deleted.
+			go c.cleanUpAllAssignedIdentitiesOnNode(nodeName, nodeTrackList)
```
Comment (kkmsft) [inline pkg/mic/mic.go]: Is there a reason for us to perform this async ?

Comment (kkmsft) [reply pkg/mic/mic.go]: Is there anything preventing another sync loop to come and perform the same operations again ?

File: pkg/mic/mic.go [hunk 5]
```diff
@@ -874,31 +865,72 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
...
 		}
 		c.EventRecorder.Event(binding, corev1.EventTypeNormal, "binding removed",
-			fmt.Sprintf("Binding %s removed from node %s for pod %s", binding.Name, createID.Spec.NodeName, createID.Spec.Pod))
+			fmt.Sprintf("Binding %s removed from node %s for pod %s", binding.Name, deleteID.Spec.NodeName, deleteID.Spec.Pod))
 	}
+}
 
-	for _, deleteID := range nodeTrackList.assignedIDsToDelete {
-		binding := deleteID.Spec.AzureBindingRef
+func (c *Client) consolidateVMSSNodes(nodeMap map[string]trackUserAssignedMSIIds) {
```
Comment (kkmsft) [inline pkg/mic/mic.go]: Can you please add comments here on overall approach of this function.

File: pkg/mic/vmss.go (hunk 1 above)
Comment (aramase) [reply pkg/mic/vmss.go]: If the vmssID is erroneous, then we fail while trying to create the client and not perform any operations.

File: pkg/mic/mic.go (hunk 2 above)
Comment (aramase) [reply pkg/mic/mic.go]: Used only here, so didn't define as a const.

File: pkg/mic/mic.go (hunk 4 above)
Comment (aramase) [reply pkg/mic/mic.go]: We don't want the clean up to block the node processing for rest of the nodes/identities. Good point about preventing sync loop doing again. We can create a WaitGroup at sync loop and pass that for consolidating (used if there is a clean up for node not existing) and for update node identities. That way we can block until all node processing is complete.

File: pkg/mic/mic.go [hunk 6]
```diff
@@ -866,39 +857,82 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
...
+		for i := 0; i < len(vmssNodes); i++ {
+			vmssTrackList.addUserAssignedMSIIDs = append(vmssTrackList.addUserAssignedMSIIDs, nodeMap[vmssNodes[i]].addUserAssignedMSIIDs...)
+			vmssTrackList.removeUserAssignedMSIIDs = append(vmssTrackList.removeUserAssignedMSIIDs, nodeMap[vmssNodes[i]].removeUserAssignedMSIIDs...)
+			vmssTrackList.assignedIDsToCreate = append(vmssTrackList.assignedIDsToCreate, nodeMap[vmssNodes[i]].assignedIDsToCreate...)
+			vmssTrackList.assignedIDsToDelete = append(vmssTrackList.assignedIDsToDelete, nodeMap[vmssNodes[i]].assignedIDsToDelete...)
+			vmssTrackList.isvmss = true
+
+			delete(nodeMap, vmssNodes[i])
+
+			nodeMap[getVMSSName(vmssName)] = vmssTrackList
```
Comment (kkmsft) [inline pkg/mic/mic.go]: Can you please add comments which specifies that we are using nodeMap with both kind of keys.

File: pkg/mic/vmss.go (hunk 1 above)
Comment (kkmsft) [reply pkg/mic/vmss.go]: Can we please handle the error for the path.Split returning an empty string ?

File: pkg/mic/mic.go [hunk 7]
```diff
@@ -866,39 +857,82 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
...
+
+	// aggregate vmss nodes into vmss name
+	for vmssName, vmssNodes := range vmssMap {
+		if len(vmssNodes) < 1 {
+			continue
+		}
+
+		vmssTrackList := trackUserAssignedMSIIds{}
+
+		for i := 0; i < len(vmssNodes); i++ {
```
Comment (kkmsft) [inline pkg/mic/mic.go]: Why index rather than range usage ?

File: pkg/mic/mic.go [hunk 8]
```diff
@@ -866,39 +857,82 @@ func (c *Client) updateUserMSI(newAssignedIDs []aadpodid.AzureAssignedIdentity,
...
+		if err != nil && !strings.Contains(err.Error(), "not found") {
+			glog.Errorf("Unable to get node %s. Error %v", nodeName, err)
+			continue
+		}
+		if err != nil && strings.Contains(err.Error(), "not found") {
+			glog.Warningf("Unable to get node %s while updating user msis. Error %v", nodeName, err)
+			wg.Add(1)
+			// node is no longer found in the cluster, all the assigned identities that were created in this sync loop
+			// and those that already exist for this node need to be deleted.
+			go c.cleanUpAllAssignedIdentitiesOnNode(nodeName, nodeTrackList, wg)
```
Comment (kkmsft) [inline pkg/mic/mic.go]: If we use anonymous function, we don't have to pass wg to cleanUpAllAssignedIdentitiesOnNode. Is that idiomatic way to use ?

File: pkg/mic/mic.go [hunk 9]
```diff
@@ -310,8 +310,15 @@ func (c *Client) Sync(exit <-chan struct{}) {
 			c.getListOfIdsToAssign(*addList, nodeMap)
 		}
 
+		var wg sync.WaitGroup
```
Comment (kkmsft) [inline pkg/mic/mic.go]: Can you please add a comment near nodeMap make to indicate that the key can be now vmss or vm name.

File: pkg/cloudprovider/cloudprovider.go [hunk 10]
```diff
@@ -153,8 +152,8 @@ func withInspection() autorest.PrepareDecorator {
 }
 
 // GetUserMSIs will return a list of all identities on the node
```
Comment (kkmsft) [inline pkg/cloudprovider/cloudprovider.go]: nit:s/Node/Node or VMSS depending on the value of isVMSS

File: pkg/mic/mic.go (hunk 8 above)
Comment (aramase) [reply pkg/mic/mic.go]: I'm passing the same wg for consolidating vmss node (which perform the cleanup if reqd) and update node deps. That way all our processing is done async.

**Pull request merged**